tools/bumper/bumper.go

```diff
@@ -0,0 +1,180 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/pkg/errors"
+)
+
+var logger *log.Logger
+
+type inputParams struct {
+	componentsConfigPath string
+	gitToken             string
+}
+
+const (
+	updatePolicyStatic = "static"
+	updatePolicyTagged = "tagged"
+	updatePolicyLatest = "latest"
+)
+
+func main() {
+	logger = initLog()
+	logger.Printf("~~~~~~~~Bumper Script~~~~~~~~")
+
+	inputArgs := inputParams{}
+	initFlags(&inputArgs)
+
+	githubApi, err := newGithubApi(inputArgs.gitToken)
+	if err != nil {
+		exitWithError(errors.Wrap(err, "Failed to create github api instance"))
+	}
+
+	logger.Printf("Parsing %s", inputArgs.componentsConfigPath)
+	componentsConfig, err := parseComponentsYaml(inputArgs.componentsConfigPath)
+	if err != nil {
+		exitWithError(errors.Wrap(err, "Failed to parse components yaml"))
+	}
+
+	for componentName, component := range componentsConfig.Components {
+		logger.Printf("~~Checking if %s needs bumping~~", componentName)
+
+		err = printCurrentComponentParams(component)
+		if err != nil {
+			exitWithError(errors.Wrapf(err, "Failed to print component %s", componentName))
+		}
+
+		gitComponent, err := newGitComponent(githubApi, componentName, &component)
+		if err != nil {
+			exitWithError(errors.Wrapf(err, "Failed to clone %s", componentName))
+		}
+		defer os.RemoveAll(gitComponent.gitRepo.localDir)
+
+		currentReleaseTag, err := gitComponent.getCurrentReleaseTag()
+		if err != nil {
+			exitWithError(errors.Wrapf(err, "Failed to get current release version tag from %s", componentName))
+		}
+
+		updatedReleaseTag, updatedReleaseCommit, err := gitComponent.getUpdatedReleaseInfo()
+		if err != nil {
+			exitWithError(errors.Wrapf(err, "Failed to get latest release version tag from %s", componentName))
+		}
+
+		bumpNeeded, err := isBumpNeeded(currentReleaseTag, updatedReleaseTag, component.Updatepolicy)
```

Comment:
We have the component fields inside `gitComponent` so we can call `getCurrentRelease` and `getUpdateReleaseInfo` inside `isBumpNeeded` and `isBumpNeeded` can be call `calculateBump` and return a `*bumpInfo` struct with `updatedReleaseCommit` and `updateReleaseTag` if is nill there is no bump needed else you use the fields there.

Reply:
I don't see the advantage in putting everything inside `isBumpNeeded` (/`calculateBump`).
Right now I can test `isBumpNeeded()` with fabricated data to test only the bump decision logic.
IMO putting `getCurrentReleaseTag` and `getUpdatedReleaseInfo` complicates testing without real value.

What do you think?

Reply:
oh.. forgot you're not available. @phoracek ?

Reply:
I tend to agree with Ram. It is better to have flat functions dependency tree instead of nesting too much. It helps with testing and keeps my brain RAM sane.